### PR TITLE
impl(bigquery): InsertJob api prep work - Part 2

### PR DESCRIPTION
**Pull request by jsrinnn**
This PR adds the statistical information to the Job custom resource. Includes the following changes

1) New definitions for JobStatistics and JobQueryStatistics and related classes.

2) Json Parsers and unit tests for the above.

In the followup PR I'll add the modifcation to Job and JobConfiguration classes to include the above and related unit tests.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784)

### CODE REVIEW

File: google/cloud/bigquery/v2/minimal/internal/job_query_stats.cc [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+    p.avg_previous_execution_time = std::chrono::milliseconds(millis);
+  }
+  if (j.contains("stage_performance_standalone_insights"))
+    j.at("stage_performance_standalone_insights")
+        .get_to(p.stage_performance_standalone_insights);
+  if (j.contains("stage_performance_change_insights"))
+    j.at("stage_performance_change_insights")
+        .get_to(p.stage_performance_change_insights);
+}
+void to_json(nlohmann::json& j, JobQueryStatistics const& q) {
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/job_query_stats.cc:354]: nit: blank line before function definition

File: google/cloud/bigquery/v2/minimal/internal/job_stats.cc [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  if (j.contains("reservation_id"))
+    j.at("reservation_id").get_to(s.reservation_id);
+  if (j.contains("row_level_security_applied"))
+    j.at("row_level_security_applied").get_to(s.row_level_security_applied);
+  if (j.contains("data_masking_applied"))
+    j.at("data_masking_applied").get_to(s.data_masking_applied);
+  if (j.contains("completion_ratio"))
+    j.at("completion_ratio").get_to(s.completion_ratio);
+  if (j.contains("start_time")) {
+    std::int64_t millis;
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/job_stats.cc:86]: I think you should consider a clean up PR where you refactor these conversions to and from milliseconds to something more like:

```cc
  s.start_time = FromMillisecondField(j, "start_time");
```

or even

```cc
  FromMillisecondField(s.start_time, j, "start_time");
```

where the helper would be something like:

```cc 
void FromMillisecondField(std::chrono::milliseconds& field, nlohman::json const& j, char const* name) {
  auto const l = j.find(name);
  if (l == j.end()) return;
  std::int64_t m;
  l->get_to(m);
  field = std::chrono::milliseconds(m);
}
```

File: google/cloud/bigquery/v2/minimal/internal/job_stats.cc [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+bool operator==(ScriptStackFrame const& lhs, ScriptStackFrame const& rhs) {
+  return lhs.end_column == rhs.end_column && lhs.end_line == rhs.end_line &&
+         lhs.procedure_id == rhs.procedure_id &&
+         lhs.start_column == rhs.start_column &&
+         lhs.start_line == rhs.start_line && lhs.text == rhs.text;
+}
+
+bool operator==(ScriptStatistics const& lhs, ScriptStatistics const& rhs) {
+  bool eq = lhs.evaluation_kind == rhs.evaluation_kind;
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/job_stats.cc:126]: nit: `auto const eq` or `bool const eq` if you must.

**Review by coryan — APPROVED** [review]
A few nits.  Please do break these large changes down.  It is too easy to get distracted on large changes and miss things.

**Review by jsrinnn — COMMENTED** [review]
Ack. 

My apologies. I clubbed these as they were following the same patterns as previous PRs and also I already have 4 PRs as follow up to this one for the custom classes. There are a lot of code changes and I was trying to speed it up but I do recognize the burden on the reviewer. Will try to be more mindful.

Appreciate the review. Thank You!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784#-:-NXFtfr7-FzkPZ-8ZXzN:b-xhdu1s)* status: 0 of 15 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/bigquery/v2/minimal/internal/job_query_stats.cc` line 354 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784#gh-1219572794:-NXFwKTBAGwMk5-CFMQG:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eee0cf1e17572d0feda374472b14f1ed514d7ee5/google/cloud/bigquery/v2/minimal/internal/job_query_stats.cc#L354)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: blank line before function definition
</blockquote></details>

Done
___
*[`google/cloud/bigquery/v2/minimal/internal/job_stats.cc` line 86 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784#gh-1219585795:-NXFwYEq0VML_jdaMlNz:b-uhwoyt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eee0cf1e17572d0feda374472b14f1ed514d7ee5/google/cloud/bigquery/v2/minimal/internal/job_stats.cc#L86)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think you should consider a clean up PR where you refactor these conversions to and from milliseconds to something more like:

```cc
  s.start_time = FromMillisecondField(j, "start_time");
```

or even

```cc
  FromMillisecondField(s.start_time, j, "start_time");
```

where the helper would be something like:

```cc 
void FromMillisecondField(std::chrono::milliseconds& field, nlohman::json const& j, char const* name) {
  auto const l = j.find(name);
  if (l == j.end()) return;
  std::int64_t m;
  l->get_to(m);
  field = std::chrono::milliseconds(m);
}
```


</blockquote></details>

Ack. Will do this in a followup PR after this one.

Added a Github issue to track this: https://github.com/googleapis/google-cloud-cpp/issues/11786 
and a TODO here.
___
*[`google/cloud/bigquery/v2/minimal/internal/job_stats.cc` line 126 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784#gh-1219586902:-NXFzl7j8K5JUJEqscfW:bb6un1j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/eee0cf1e17572d0feda374472b14f1ed514d7ee5/google/cloud/bigquery/v2/minimal/internal/job_stats.cc#L126)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: `auto const eq` or `bool const eq` if you must.
</blockquote></details>

Done

**Review by jsrinnn — COMMENTED** [review]
Addressed code review feedback. Thanks!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784#-:-NXG7rL33QC8sAUkQaFI:b-8e44jq)* status: 0 of 15 files reviewed, 1 unresolved discussion (waiting on @coryan)

**Review by coryan — APPROVED** [review]
> Reviewed 13 of 15 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11784)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @jsrinnn)

**Pull request merged**